src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubRDDIterator.java

```diff
@@ -0,0 +1,169 @@
+ */
+package io.cdap.plugin.gcp.publisher.source;
+
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.stub.GrpcSubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStubSettings;
+import com.google.pubsub.v1.AcknowledgeRequest;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.ReceivedMessage;
+import org.apache.spark.TaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.Iterator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Iterator for PubSub RDD.
+ * Fetches a fixed amount of messages at a time and acknowledges them.
+ * Finishes when the acknowledged messages are completed and the batch time limit is met.
+ * Returns immediately if there are no more messages to read.
+ */
+public class PubSubRDDIterator implements Iterator<PubSubMessage> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PubSubRDDIterator.class);
+  private static final int MAX_MESSAGES = 1000;
+
+  private final long startTime;
+  private final PubSubSubscriberConfig config;
+  private final TaskContext context;
+  private final long batchDuration;
+  private final Credentials credentials;
+  private final String subscriptionFormatted;
+  private PullRequest pullRequest;
+  private SubscriberStub subscriber;
+  private boolean autoAcknowledge;
+  private Queue<ReceivedMessage> receivedMessages;
+  private List<String> ackIds;
+  private long messageCount;
+
+  public PubSubRDDIterator(PubSubSubscriberConfig config, TaskContext context, long batchDuration,
+                           boolean autoAcknowledge, Credentials credentials) {
+    this.config = config;
+    this.context = context;
+    this.batchDuration = batchDuration;
+    this.credentials = credentials;
+    this.startTime = System.currentTimeMillis();
+    this.autoAcknowledge = autoAcknowledge;
+    subscriptionFormatted = ProjectSubscriptionName.format(this.config.getProject(), this.config.getSubscription());
+    receivedMessages = new ConcurrentLinkedDeque<>();
+    ackIds = new ArrayList<>();
+  }
+
+  @Override
+  public boolean hasNext() {
+    // Complete processing of acknowledged messages before finishing the batch.
+    if (!receivedMessages.isEmpty()) {
+      return true;
+    }
+
+    long currentTimeMillis = System.currentTimeMillis();
+    if (currentTimeMillis >= (startTime + batchDuration)) {
+      LOG.debug("Time exceeded for batch. Total time is {} millis. Total messages returned is {} .",
+                currentTimeMillis - startTime, messageCount);
+      return false;
```

Comment:
since we've already acked the messages, doesn't this mean anything in the receivedMessages queue will get skipped? should we always go through the list, even if we're past the duration?

Reply:
The check above this will make sure that if queue has pending messages hasNext() will return true. Only when queue is empty, the time is checked.